### ISSUE TITLE
Disabling "Upgrade connections to HTTPS" works on desktop, but not on Android.

### ISSUE THREAD

**Issue by rabinnh** [brave/brave-browser#38197]
## Description 
Brave on Android redirects to https even when "Upgrade connections to HTTPS" setting is set to disabled,

## Steps to reproduce 

   1. Open Brave browser on Android
   2. Open settings
   3. In the "Brave Shields and Privacy" find the "Upgrade connections to HTTPS"
   4. Set it to disable

## Actual result 
Accessing a bookmarked site using the http protocol redirects to https.  Manually changing the protocol in the address bar to http and Brave still redirects to https.

## Expected result
This setting works perfectly on my desktop. Desktop Brave is synced to Android Brave (FWIW).

## Issue reproduces how often 
Every time.

## Version/Channel Information:

- Can you reproduce this issue with the current Play Store version? 
- Yes
- Can you reproduce this issue with the current Play Store Beta version? 
- Haven't installed because synced
- Can you reproduce this issue with the current Play Store Nightly version?
Haven't installed because synced
## Device details
   - Install type (ARM, x86): ARM
   - Device type (Phone, Tablet, Phablet): Google Pixel 6
   - Android version: 14
   
## Brave version
1.65.126

### Website problems only
- Does the issue resolve itself when disabling Brave Shields? No
- Does the issue resolve itself when disabling Brave Rewards? No
- Is the issue reproducible on the latest version of Chrome? Yes

**Comment by TEMP-ad** [comment]
You needed to research and understand what the **HTTPS  Upgrade** does before filing this issue, because it is clear it does something you haven't researched about.

First, the HTTPS upgrade it is not exactly about the HTTP changing the url to HTTPS always, it is part of it, but not exactly what it is, this is also affected because of the way the _https upgrade_ feature was replaced recently by Brave, it actually avoids trying to make a URL https when it is http if it cause navigation errors, so that's pretty important because an http will not go https if it breaks anything, if the site for some reasons tries to upgrade and causes an error, there is an _exception list_ so sites can be excluded from that change of URL.

You have to remember the 'httpS' status is granted because of a valid SSL certificate, without a valid certificate, even if https exists, it will be 'not secured' and in read.

So in the end, you don't upgrade from http to https only because you have the option on, and that means you don't stop redirecting from http to https only because you turn it off either, it is the logic behind the https upgrade feature, so the web doesn't break as much because of it.

Now, you have to understand that Websites will most likely redirect http to https automatically, so doesn't matter if the option is on or off, each domain pretty much has the power to do that and they do it often, because they don't need to give you a http website that works fine with a valid SSL certificate and all.

Some domains will not do that though, and you can visit both http and https and they will work fine.  

So you are testing it wrong by using any random website, because mobile has a different user agent and you don't know if that has something to do with the difference between PC and Phone. Also, Private windows will not use the global _never upgrade_ BTW, it is always set to **Standard / Upgrade connections to HTTPS**. so if you are testing in private windows, that will also affect things, you would have to do it in normal windows so the _never_ is respected.

### Examples to test this and show you how it is working as it is intended.

1. if you go to https://www.httpvshttps.com/ and you click on http, you will keep getting 'redirected' to the https test, if you set the Shields to **Never / Don't upgrade HTTPS connections**, you will be able to test the http one. Standard or  **Strict / Only connect with HTTPS**  will go to https.
2. https://badssl.com/ with this website you can test how the HTTP works, pretty much if you set it to **Strict** it will give you a message:
   > The connection to http.badssl.com is not secure
   > You are seeing this warning because this site does not support HTTPS.
   > Continue to site or Go Back
   
   In this case, that will be the only difference, because Never and Standard will still connect to the _Not Secured_ HTTP and show you the red screen. If you try to go to HTTPS, it will go back to HTTP as well.
3. http://toptanparcakontor.com I got this URL from the exception list, this website will not redirect to https, but it gets redirected if you set the option to Strict, Standard and Never will not do it, and this is because since it is in the exception list, the Standard will not upgrade it as the list says, because if you remove the website from the exception list, it will get upgraded on Standard, so the exception list is only for Standard mode, and since there is no 'error' navigating to HTTPS then it upgrades it.
   Of course like I said, if the SSL is not valid and the website is insecure, the Strict will just display the screen asking if you want to go to website anyway or go back.
4. https://trckmaniac.com/ another website of the exception list, this will only work in Never or Standard, because Strict will make it https and https will not display the content of the page and only say _Tracking domain is ready_ . It is a valid HTTPS page so no error or anything, but it will not display the correct page. Of course, if you remove the site from the exception list, it will not break in Standard too.
5. http://volokit2.com/ this website will not work if you switch it to HTTPS, this website will not get upgraded to HTTPS in Strict mode, you will get the error if you want to continue, but that's it, it will still be a http, because the feature will detect the page doesn't work if https.

So as you can see, this is the way the feature works, it is not just a simple http to https and that's it, this is also why the exception list exist and the different situation in websites and why running strict might not work in some domains where http is the good one but it still offers the https version.

So it works in Android, I tested it and could do the HTTP test fine. You just need to understand what this feature does exactly. But in the end, most website will redirect automatically to https when you go to a http, that is the correct way of doing things, so you will not see too many 'https' upgrades being done.

**Comment by diracdeltas** [comment]
cc @SergeyZhukovsky

**Comment by SergeyZhukovsky** [comment]
thanks @rabinnh and thanks for the ping @diracdeltas. I'm going to take a look at it.

**Comment by SergeyZhukovsky** [comment]
Everything works fine on my side. I tried on http://https-everywhere.badssl.com/. I attached a video below.
@rabinnh what I think happened on your side is that you interacted with the website you are trying with Brave Shields. Once you touch Brave Shields(a lion's head on the URL bar) the site shields settings are going to override global settings(that's the main purpose of it). So it saved all shields settings for the website you use and HTTPS upgrades are on there.
[Screen_recording_20240514_144425.webm](https://github.com/brave/brave-browser/assets/12011303/7ef06cf2-7a3f-42ef-83d0-589b49f67eb7)

**Comment by SergeyZhukovsky** [comment]
1. @rabinnh does http://https-everywhere.badssl.com/ got upgraded to https when global HTTPS upgrades are turned off?

**Comment by wrobelda** [comment]
I think I run into this issue on Desktop! If I test http://https-everywhere.badssl.com/  and disable Upgrade HTTP for it, Brave respects that and indeed does *not* redirect to https. However, I have a local website which does *not* respond to HTTPS at all and disabling Upgrade HTTPS for it has no effect – Brave still insists on redirecting to HTTPS. I wonder if this has anything to do with the fact that the HTTPS version of that website responds with 404? Is it possible that per-site settings are only saved/respected for sites that actually properly loaded in the first place?

And, for the record, I have same exact issue on Android.

**Comment by SergeyZhukovsky** [comment]
@wrobelda could you reproduce that in Chrome as well?

**Comment by wrobelda** [comment]
> [@wrobelda](https://github.com/wrobelda) could you reproduce that in Chrome as well?

I am on aarch64, there's unfortunately no Chrome build available. in Chromium it just works, though, but I don't think they force HTTPS redirection at all.

**Comment by SergeyZhukovsky** [comment]
@ShivanKaul any thoughts on that https://github.com/brave/brave-browser/issues/38197#issuecomment-4019792574?

**Comment by ShivanKaul** [comment]
> However, I have a local website which does not respond to HTTPS at all and disabling Upgrade HTTPS for it has no effect – Brave still insists on redirecting to HTTPS.

By "local website" you mean something on localhost? That doesn't match expected behavior and I can't repro. Can you send a screen recording?

**Comment by wrobelda** [comment]
> > However, I have a local website which does not respond to HTTPS at all and disabling Upgrade HTTPS for it has no effect – Brave still insists on redirecting to HTTPS.
> 
> By "local website" you mean something on localhost? That doesn't match expected behavior and I can't repro. Can you send a screen recording?

No, local as in self-hosted on intranet but accessed with FQDN.  Let me get back to you on recording.

EDIT: Recording attached.

https://github.com/user-attachments/assets/720cbb3b-383c-4029-a607-160cb2aa6246

Note that the page loads fine if opened with http://valetudo (hostname only), which is expected since redirection is not enforced on hostnames.

**Comment by 0xAnakin** [comment]
This issue remains unresolved when previewing a local dev website from Brave android
